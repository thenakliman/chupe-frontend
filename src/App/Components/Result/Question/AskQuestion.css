.input-field-for-question {
}

form {
  text-align: center;
}

p {
 padding: 1rem;
 display: flex;
}

label{
  width: 20rem;
}

input {
  width: 45rem;
  border-radius: 0px;
  border: 1px solid rgb(180, 180, 180);
  height: 1.5rem;
}

textarea {
  border: 1px solid rgb(180, 180, 180);
  width: 45rem;
  height: 20rem;
  word-wrap: break-word;
}

select {
    padding: 0.3rem 1.5rem;
}

.ask-question-submit-button {
      border-radius: 0.25rem;
      border: 1px solid #28a745;
      background-color: #28a745;
      padding: 0.8rem;
      margin: 5rem;
      color: white;
      background-image: linear-gradient(-180deg, #34d058 0%, #28a745 90%);
}
