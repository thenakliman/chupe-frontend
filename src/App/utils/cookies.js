import Cookies from 'universal-cookie';
import {verify, decode} from 'jsonwebtoken';

const TOKEN_COOKIE_KEY = 'token';
const TOKEN_EXPIRY_COOKIE_KEY = 'expiryTime';
const TOKEN_USERNAME_KEY = 'username';
const audience = 'chupe-frontend';
const issuer = 'chupe';
export let cookies = new Cookies();
const Config = require('Config');

const setCookie = (key, value) => {
  cookies.set(key, value);
};

export const setCookies = (token) => {
  verify(token,
         Config.signingKey,
         {
            algorithms: ['HS512'],
            audience: audience,
            issuer: issuer,
            ignoreExpiration: false,
         },
         (e) => console.log(e)
  );

  const decodedToken = decode(token);
  setToken(token);
  setTokenExpiryTime(decodedToken.exp);
  setUsername(decodedToken.username);
};

const setToken = (token) => {
  setCookie(TOKEN_COOKIE_KEY, token);
};

const setTokenExpiryTime = (expiryTime) => {
  setCookie(TOKEN_EXPIRY_COOKIE_KEY, expiryTime);
};

const setUsername = (userName) => {
  setCookie(TOKEN_USERNAME_KEY, userName);
};

export const getUsername = () => {
  return cookies.get(TOKEN_USERNAME_KEY);
};

export const getToken = () => {
  if (!cookies.get(TOKEN_COOKIE_KEY) ||
      !cookies.get(TOKEN_EXPIRY_COOKIE_KEY) ||
      cookies.get(TOKEN_EXPIRY_COOKIE_KEY) < Date.now()) {
    return null;
  }

  return cookies.get(TOKEN_COOKIE_KEY);
};
