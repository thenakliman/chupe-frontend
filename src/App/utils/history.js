import {createHashHistory} from 'history';

export let history = createHashHistory();
