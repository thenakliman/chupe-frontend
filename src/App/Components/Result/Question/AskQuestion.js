import React from 'react'; // eslint-disable-line no-unused-vars
import propTypes from 'prop-types';

require('./AskQuestion.css');


/** AskQuestion component with input field having for asking
* question.
*/
export class AskQuestion extends React.Component {
    /** Constructor defining the default values for the fields.
    * @param {Object} props, properties for the component.
    */
    constructor(props) {
        super(props);
        this.state = {
            question: '',
            description: '',
            owner: '',
            assignedTo: '',
        };

        this.handleQuestionSummary = this.handleQuestionSummary.bind(this);
        this.handleQuestionDescription = this
            .handleQuestionDescription.bind(this);

        this.handleOwnerChange = this.handleOwnerChange.bind(this);
        this.handleAssignedToChange = this.handleAssignedToChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }
    /** Get all questions on component mount */
    componentDidMount() {
         this.props.getQuestions();
    }

    /** Handle changes on the question fields
    * @param {object} event containing new modified value
    */
    handleQuestionSummary(event) {
        this.setState({question: event.target.value});
    }

    /** Handle changes on the question description fields
    * @param {object} event containing new modified value
    */
    handleQuestionDescription(event) {
        this.setState({description: event.target.value});
    }

    /** Handle changes on the question owner fields
    * @param {object} event containing new modified value
    */
    handleOwnerChange(event) {
        this.setState({owner: event.target.value});
    }

    /** Handle changes on the question assigned to fields
    * @param {object} event containing new modified value
    */
    handleAssignedToChange(event) {
        this.setState({assignedTo: event.target.value});
    }

    /** Handles submit of the form
    * @param {object} event containing new modified value
    */
    handleSubmit(event) {
        event.preventDefault();
        this.props.askQuestion(this.state);
    }
    /** Returns input field in form with a button.
    *
    * @return {object} Component for asking question.
    */
    render() {
        return (
            <div id='ask-question-div' className='input-field-for-question'>
                <form id="ask-question-form" onSubmit = {this.handleSubmit}>
                    <input type="text"
                        id="ask-question-input-field"
                        value={this.state.value}
                        onChange={this.handleQuestionSummary}
                    />
                    <textarea id='question-description-textarea'
                        value={this.state.value}
                        onChange={this.handleQuestionDescription}
                    />

                    <select id='question-owner'
                        value={this.state.value}
                        onChange={this.handleOwnerChange}>
                     {
                         this.props.users.map((user) =>(
                             <option key={`${user.userName}`}>
                                {user.userName}
                             </option>))
                     }
                     </select>
                    <select id='question-assigned-to'
                        value={this.state.value}
                        onChange={this.handleAssignedToChange}>
                     {
                         this.props.users.map((user) =>(
                             <option key={`${user.userName}`}>
                                {user.userName}
                             </option>))
                     }
                     </select>
                    <button id="ask-question-submit-button"
                        type="submit"
                        value="Submit">

                        Ask Question
                    </button>
                </form>
            </div>
        );
    }
}

AskQuestion.propTypes = {
    getQuestions: propTypes.func.isRequired,
    askQuestion: propTypes.func.isRequired,
    users: propTypes.arrayOf(propTypes.object).isRequired,
};
