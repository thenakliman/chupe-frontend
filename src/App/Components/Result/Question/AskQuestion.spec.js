import React from 'react';// eslint-disable-line no-unused-vars
import {AskQuestion} from './AskQuestion';
import {shallow, mount} from 'enzyme';
import toJson from 'enzyme-to-json';
import * as History from '../../../utils/history';

describe('Question Result component snapshot', () => {
    it('should match the snapshot', () => {
        const users = [{'userName': 'user1'}, {'userName': 'user2'}];
        const wrapper = shallow(
            <AskQuestion
                askQuestion={()=>{}}
                users={users}
            />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});

describe('Question result', () => {
  let initialState;

    beforeEach(() => {
        initialState = {
          question: 'q1',
          description: 'd1',
          owner: 'o1',
          assignedTo: 'a1',
        };
    });

    it('Should display input text field', () => {
        const users = [{'userName': 'user1'}, {'userName': 'user2'}];
        const wrapper = shallow(
            <AskQuestion
                askQuestion={()=>{}}
                users={users} />);
        expect(
            wrapper.find('#ask-question-input-field')
                .get(0)
                .props
                .id
        ).toEqual('ask-question-input-field');
    });

    it('Should call ask method on click', () => {
        const users = [{'userName': 'user1'}, {'userName': 'user2'}];
        const askQuestion = jest.fn();
        const container = mount(
            <AskQuestion
                askQuestion={askQuestion}
                users={users}
            />);
        const props = container.find(AskQuestion).props();
        props.askQuestion();
        expect(props.askQuestion).toHaveBeenCalledWith();
    });

    it('Component should have required questions properties', () => {
        const users = [{'userName': 'user1'}, {'userName': 'user2'}];
        const wrapper = shallow(<AskQuestion
             askQuestion={()=>{}}
             users={users} />);

        expect(wrapper.state()).toEqual({
            question: '',
            description: '',
            owner: '',
            assignedTo: '',
        });
    });
    it('should not call submit if any field is empty', () => {
        const askQuestion = jest.fn();
        const users = [{'userName': 'user1'}, {'userName': 'user2'}];
        const container = mount(<AskQuestion
            askQuestion={askQuestion}
            users={users}
        />);
        const state = Object.assign({}, initialState, {question: ''});
        container.setState(state);
        container.find('#ask-question-input-field').simulate('submit');
        expect(askQuestion).not.toHaveBeenCalled();
    });

    it('should set question text field', () => {
        const askQuestion = jest.fn();
        const users = [{'userName': 'user1'}, {'userName': 'user2'}];
        const event = {target: {value: 'why oh why?'}};
        const container = mount(<AskQuestion
            askQuestion={askQuestion}
            users={users}
        />);
        const state = Object.assign({}, initialState, {question: ''});
        container.setState(state);
        container.find('#ask-question-input-field')
            .simulate('change', event);

        container.find('#ask-question-input-field').simulate('submit');
        expect(askQuestion).toHaveBeenCalledWith(
            Object.assign({}, initialState, {'question': event.target.value}));
    });

    it('should set text field', () => {
        const askQuestion = jest.fn();
        History.history = {push: jest.fn()};
        const users = [{'userName': 'user1'}, {'userName': 'user2'}];
        const event = {target: {value: 'Description'}};
        const container = mount(<AskQuestion
            askQuestion={askQuestion}
            users={users}
        />);
        const state = Object.assign({}, initialState, {description: ''});
        container.setState(state);
        container.find('#question-description-textarea')
            .simulate('change', event);

        container.find('#ask-question-submit-button').simulate('submit');
        expect(askQuestion).toHaveBeenCalledWith(
            Object.assign(
                {},
                initialState,
                {'description': event.target.value}));
        expect(History.history.push).toHaveBeenCalledWith('/questions');
    });

    it('should select owner field', () => {
        const askQuestion = jest.fn();
        History.history = {push: jest.fn()};
        const event = {target: {value: 'user1'}};
        const users = [{'userName': 'user1'}, {'userName': 'user2'}];
        const container = mount(<AskQuestion
            askQuestion={askQuestion}
            users={users}
        />);
        const state = Object.assign({}, initialState, {owner: ''});
        container.setState(state);
        container.find('#question-owner').simulate('change', event);
        container.find('#ask-question-submit-button').simulate('submit');
        expect(askQuestion).toHaveBeenCalledWith(
            Object.assign({}, initialState, {'owner': event.target.value}));
        expect(History.history.push).toHaveBeenCalledWith('/questions');
    });

    it('should select assigned to field', () => {
        const askQuestion = jest.fn();
        History.history = {push: jest.fn()};
        const event = {target: {value: 'user1234'}};
        const users = [{'userName': 'user1234'}, {'userName': 'user2'}];
        const container = mount(<AskQuestion
            askQuestion={askQuestion}
            users={users}
        />);
        const state = Object.assign({}, initialState, {assignedTo: ''});
        container.setState(state);
        container.find('#question-assigned-to').simulate('change', event);
        container.find('#ask-question-submit-button').simulate('submit');
        expect(askQuestion).toHaveBeenCalledWith(
            Object.assign(
                {},
                initialState,
                {'assignedTo': event.target.value}));
        expect(History.history.push).toHaveBeenCalledWith('/questions');
    });
});
