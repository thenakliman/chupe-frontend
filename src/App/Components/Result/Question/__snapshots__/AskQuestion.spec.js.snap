// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question Result component snapshot should match the snapshot 1`] = `
<div
  className="input-field-for-question"
  id="ask-question-div"
>
  <form
    id="ask-question-form"
    onSubmit={[Function]}
  >
    <input
      id="ask-question-input-field"
      onChange={[Function]}
      type="text"
    />
    <textarea
      id="question-description-textarea"
      onChange={[Function]}
    />
    <select
      id="question-owner"
      onChange={[Function]}
    >
      <option
        key="user1"
      >
        user1
      </option>
      <option
        key="user2"
      >
        user2
      </option>
    </select>
    <select
      id="question-assigned-to"
      onChange={[Function]}
    >
      <option
        key="user1"
      >
        user1
      </option>
      <option
        key="user2"
      >
        user2
      </option>
    </select>
    <button
      id="ask-question-submit-button"
      type="submit"
      value="Submit"
    >
      Ask Question
    </button>
  </form>
</div>
`;
